### PR TITLE
Support XCFrameworks with missing architectures in the Info.plist

### PR DESCRIPTION
**Pull request by iainsmith**
### Short description 📝

This is a very small UX improvement for the edge case where users have usable XCFrameworks, but the Info.plist references missing architectures.

Handle the case where users have deleted a binary from an xcframework to save binary size (e.g maccatalyst) but
left the reference in the Info.plist. Provide a warning to let users know their xcframework is missing an
architecture.

This is typically a recoverable error as long as there is 1 valid architecture tuist can use.

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] ~(having issues with Fourier locally)~ The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] (No user facing change) In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

File: Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift [hunk 1]
```diff
@@ -89,11 +89,25 @@ public final class XCFrameworkMetadataProvider: PrecompiledMetadataProvider, XCF
 
     public func binaryPath(xcframeworkPath: AbsolutePath, libraries: [XCFrameworkInfoPlist.Library]) throws -> AbsolutePath {
         let archs: [BinaryArchitecture] = [.arm64, .x8664]
-        guard let library = libraries.first(where: { !$0.architectures.filter(archs.contains).isEmpty }) else {
+        let binaryName = xcframeworkPath.basenameWithoutExt
+
+        guard let library = libraries.first(where: {
+            let validArchitectures = !$0.architectures.filter(archs.contains).isEmpty
+            guard validArchitectures, let binaryPath = try? path(for: $0, binaryName: binaryName, xcframeworkPath: xcframeworkPath) else { return false }
+            guard localFileSystem.exists(binaryPath) else {
```
Comment (kwridan) [inline Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift:97]: minor: Can we use `FileHandler` instead? this is the main component used for file operations throughout Tuist which has corresponding mocks

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,7 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 - The `tuist dependencies` command generates a `graph.json` file for the `Carthage` dependencies. [#3043](https://github.com/tuist/tuist/pull/3043) by [@laxmorek](https://github.com/laxmorek)
 - Add --skip-ui-tests parameter to tuist test command [#2832](https://github.com/tuist/tuist/pull/2832) by [@mollyIV](https://github.com/mollyIV).
 - Add `disableBundleAccessors` generation option which disables generating Bundle extensions [#3088](https://github.com/tuist/tuist/pull/3088) by [@wojciech-kulik](https://github.com/wojciech-kulik).
+- Support XCFrameworks with missing architectures []() by [@iainsmith](https://github.com/iainsmith).
```
Comment (kwridan) [inline CHANGELOG.md:13]: This can now be updated with the PR # 🙂

**Review by kwridan — APPROVED** [review]
Thank you for this contribution @iainsmith !

Sorry for the hassle, mind running:

```sh
./fourier format swift --fix 
```

To run the code formatter _(it runs a bundled version of `swiftformat`)_.

**Comment by kwridan** [comment]
@all-contributors add @iainsmith for code

File: Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift [hunk 3]
```diff
@@ -89,11 +89,25 @@ public final class XCFrameworkMetadataProvider: PrecompiledMetadataProvider, XCF
 
     public func binaryPath(xcframeworkPath: AbsolutePath, libraries: [XCFrameworkInfoPlist.Library]) throws -> AbsolutePath {
         let archs: [BinaryArchitecture] = [.arm64, .x8664]
-        guard let library = libraries.first(where: { !$0.architectures.filter(archs.contains).isEmpty }) else {
+        let binaryName = xcframeworkPath.basenameWithoutExt
+
+        guard let library = libraries.first(where: {
+            let validArchitectures = !$0.architectures.filter(archs.contains).isEmpty
```
Comment (fortmarek) [inline Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift:95]: maybe `hasValidArchitectures` is more apt name here?

File: Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift [hunk 4]
```diff
@@ -89,11 +89,25 @@ public final class XCFrameworkMetadataProvider: PrecompiledMetadataProvider, XCF
...
+            guard validArchitectures, let binaryPath = try? path(for: $0, binaryName: binaryName, xcframeworkPath: xcframeworkPath) else { return false }
+            guard localFileSystem.exists(binaryPath) else {
+                let relativePath  = binaryPath.components.suffix(3).joined(separator: "/")
+                logger.warning("\(xcframeworkPath.basename) is missing architecture \(relativePath) defined in the Info.plist")
```
Comment (fortmarek) [inline Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift:99]: I think it'd be nice to test this in the unit test via `XCTAssertPrinterOutputContains`

File: Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift [hunk 5]
```diff
@@ -89,11 +89,25 @@ public final class XCFrameworkMetadataProvider: PrecompiledMetadataProvider, XCF
...
     public func binaryPath(xcframeworkPath: AbsolutePath, libraries: [XCFrameworkInfoPlist.Library]) throws -> AbsolutePath {
         let archs: [BinaryArchitecture] = [.arm64, .x8664]
-        guard let library = libraries.first(where: { !$0.architectures.filter(archs.contains).isEmpty }) else {
+        let binaryName = xcframeworkPath.basenameWithoutExt
+
+        guard let library = libraries.first(where: {
+            let validArchitectures = !$0.architectures.filter(archs.contains).isEmpty
+            guard validArchitectures, let binaryPath = try? path(for: $0, binaryName: binaryName, xcframeworkPath: xcframeworkPath) else { return false }
+            guard localFileSystem.exists(binaryPath) else {
+                let relativePath  = binaryPath.components.suffix(3).joined(separator: "/")
```
Comment (fortmarek) [inline Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift:98]: maybe a comment that describes what the path is expected to be and why it makes sense to reference the last three components?

**Review by fortmarek — APPROVED** [review]
This is small, but nifty improvement, thanks for doing this 🖤

**Comment by iainsmith** [comment]
> Thank you for this contribution @iainsmith !
> 
> 
> 
> Sorry for the hassle, mind running:
> 
> 
> 
> ```sh
> 
> ./fourier format swift --fix 
> 
> ```
> 
> 
> 
> To run the code formatter _(it runs a bundled version of `swiftformat`)_.

No hassle at all.

File: Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift (hunk 4 above)
Comment (iainsmith) [reply Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift]: Good shout, I hadn't seen those helpers. Added in the latest push.

File: Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift (hunk 5 above)
Comment (iainsmith) [reply Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift]: Added a comment for this 👍

File: CHANGELOG.md (hunk 2 above)
Comment (iainsmith) [reply CHANGELOG.md]: Done

**Comment by iainsmith** [comment]
Thanks @fortmarek & @kwridan. Should be ready for another review.

**Review by fortmarek — APPROVED** [review]
LGTM once the CI is green ✅

**Pull request merged**